### ISSUE TITLE
Training NER with custom data. Data may be group of words

### ISSUE THREAD

**Issue by Prashantietedelhi** [explosion/spaCy#1363]
Hi 

I have the skills list like Java, Painter, Teacher, Account Director, Ruby on Rails..... etc.

So in my case the data may be group of words (ex: Ruby on Rails)

How should I create the custom NER model for these data. I don't want to use matcher because there are list of skills in the world.

**Comment by Prashantietedelhi** [comment]
I have tried with the solution mentioned in https://github.com/explosion/spaCy/blob/master/examples/training/train_new_entity_type.py

I am creating the data as:
                              (u'Blue Coat', [(0, 9, u'SKILL')])

But the solution does not work.

**Comment by adam-ra** [comment]
I confirm, I've tested both the quoted example and the code below, both return no entities :(

The code below is basically taken from the docs:

```
import spacy
import random
from spacy.gold import GoldParse
from spacy.language import EntityRecognizer

train_data = [
    ('Who is Chaka Khan?', [(7, 17, 'PERSON')]),
    ('I like London and Berlin.', [(7, 13, 'LOC'), (18, 24, 'LOC')])
]

nlp = spacy.load('en_depent_web_md', entity=False)
ner = EntityRecognizer(nlp.vocab, entity_types=['PERSON', 'LOC'])

for itn in range(1000):
    random.shuffle(train_data)
    for raw_text, entity_offsets in train_data:
        doc = nlp.make_doc(raw_text)
        gold = GoldParse(doc, entities=entity_offsets)

        nlp.tagger(doc)
        ner.update(doc, gold)
ner.model.end_training()

doc = nlp.make_doc('I like London and Berlin.')
nlp.tagger(doc)
print(ner(doc))
```

**Comment by dtsukiyama** [comment]
I built a skills based NER for work. I used spaCy v2.0.0 alpha as well as earlier versions of spaCy, the code you linked to did work for me on my initial attempt to build a skills based NER. However if the data you displayed is an example of your training data, then that does not look correct. I used annotated sentences like the example shows:

```
(
            "Horses are too tall and they pretend to care about your feelings",
            [(0, 6, 'ANIMAL')],
)
```
I also used over 100,000 training samples. 

However, recently this past week retraining my NERs has not gone as anticipated, so perhaps there is something wrong, I have been wondering if it is a problem with me or with spaCy. 

First retraining my NER in spaCy 1.9 seemed to overfit significantly, the NER was not longer able to generalize to sentences. I would recognize the skill only if that was the only text fed into it. Using spaCy 2.0.0 alpha also did not work, as retraining my NER resulted in it not returning anything. Additionally it took a very long time to train, I had 40000 training examples, and it took 23 hours. So I am a bit at a loss here. Things were working fine, given the same workflow, two weeks ago.

**Comment by Prashantietedelhi** [comment]
I have a list of skills 

e.x
[ u'jasper reports',
  u'jasperreports',
  u'Java/J2EE',
  u'java3d',
  u'JavaBeans',
  u'javacard',
  u'javacc',
  u'JavaDoc',
  u'JavaMail',
  u'javase',
  u'jaws',]

and I wants to tag it as a 'SKILL'

When someone write "I am a java developer".
The NER system should give
Java ---> SKILL

**Comment by dtsukiyama** [comment]
As I understand it, that won't work. In addition to the list of skills you will need annotated sentences that contain your skills.

**Comment by Prashantietedelhi** [comment]
What could be the other option?

**Comment by dtsukiyama** [comment]
You can use matching. This should work:

```
# build matcher: dtsukiyama
# spacy 1.9

from spacy.matcher import Matcher
from spacy.attrs import IS_PUNCT, LOWER

nlp = spacy.load('en')

matcher = Matcher(nlp.vocab)

def skillPattern(skill):
    pattern = []
    for b in skill.split():
        pattern.append({'LOWER':b})  
    return pattern

def buildPatterns(skills):
    pattern = []
    for skill in skills:
        pattern.append(skillPattern(skill))
    return list(zip(skills, pattern))

def buildMatcher(patterns):
    matcher = Matcher(nlp.vocab)
    for pattern in patterns:
        matcher.add_pattern(pattern[0], pattern[1])
    return matcher

def skillMatcher(matcher, text):
    skills = []
    doc = nlp(unicode(text.lower()))
    matches = matcher(doc)
    for b in matches:
        match_id, _, start, end = b
        skills.append(doc[start : end])
    return list(set([str(b) for b in skills]))
```

Your list of skills doesn't actually include java, so it should be included for matching. Also you mixed lower cased words with uppercase words, but your example, "I am a java developer," has "java" in lowercase. To make this example simpler I would just have all my examples in lowercase.

```
example = [ u'jasper reports',
u'jasperreports',
u'java/J2EE',
u'java3d',
u'javabeans',
u'javacard',
u'javacc',
u'javadoc',
u'javamail',
u'javase',
u'jaws',
u'java']
```

Now you can do the following:

```
patterns = buildPatterns(example)
skill_matcher = buildMatcher(patterns)
```

And then test it:

```
skillMatcher(skill_matcher, "I am a java developer")
['java']

skillMatcher(skill_matcher, "I am a Java developer who uses JavaBeans")
['javabeans', 'java']
```
You can have different kinds of patterns for rules-based matching https://spacy.io/docs/usage/rule-based-matching

**Comment by honnibal** [comment]
@adam-ra 
@dtsukiyama 

Try adding dropout, with the `drop` keyword argument to `nlp.parser.update()`?

There were tonnes of problems with the NER training in 1.x -- have a look through the closed threads, there's probably some relevant discussion.  That said, I'm not sure why it would've worked 2 weeks ago and not now.

About training the NER in 2.x: you probably want to be using the `spacy train` command, if you aren't already.

The current version on `develop` should train much faster than the v2.0.0a14 release. It also sets all the defaults correctly -- which would've been a significant problem. This version also lets you use the GloVe vectors as features in the model. This should make a big difference for small data sets.

Try building `develop` and doing:

     spacy download en_vectors_web_lg
     spacy train en <out_dir> <train_data> <dev_data> --vectors en_vectors_web_lg -T -P

The training and development data need to be in spaCy's (admittedly odd) json format. The easiest way to get this, if your data isn't in it already, is to put your text into `iob` format and use `spacy convert`. This also lets you artificially join your sentences, so that spaCy trains on running text. If you train on segmented sentences and test on whole documents, you may get worse performance. Concatenating the training sentences into 'pseudo paragraphs' is a simple solution. The default value of `-n 10` is a reasonable choice.

**Comment by honnibal** [comment]
@dtsukiyama Some further performance tips for spaCy 2:

- Check your processor usage. On my machines OpenBLAS is doing really badly at picking reasonable thread numbers, so I get much better performance with `export OPENBLAS_NUM_THREADS=2`
- If your CPU is at 100, you've probably got numpy linked to the reference BLAS. It'll be 5-6x faster if you can get it linked to OpenBLAS, MKL or Accelerate.

**Comment by Prashantietedelhi** [comment]
Ok got it. Thanks @dtsukiyama . For my case Matcher will be used.

**Comment by dtsukiyama** [comment]
Thanks @honnibal, very good advice!
